File: csharp/protein-translation/ProteinTranslation.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

public static class ProteinTranslation
{
    private const string Stop = "STOP";

    private static readonly IDictionary<string, string> _proteins = new Dictionary<string, string>
    {
        {"AUG", "Methionine"},
        {"UUU", "Phenylalanine"},
        {"UUC", "Phenylalanine"},
        {"UUA", "Leucine"},
        {"UUG", "Leucine"},
        {"UCU", "Serine"},
        {"UCC", "Serine"},
        {"UCA", "Serine"},
        {"UCG", "Serine"},
        {"UAU", "Tyrosine"},
        {"UAC", "Tyrosine"},
        {"UGU", "Cysteine"},
        {"UGC", "Cysteine"},
        {"UGG", "Tryptophan"},
        {"UAA", Stop},
        {"UAG", Stop},
        {"UGA", Stop}
    };

    public static string[] Proteins(string strand)
    {
        var result = new List<string>();

        foreach (var codon in Regex.Split(strand, @"(?<=\G.{3})"))
        {
            if (string.IsNullOrWhiteSpace(codon))
                continue;

            if (!_proteins.ContainsKey(codon))
                throw new KeyNotFoundException("Invalid RNA");

            var protein = _proteins[codon];
            if (protein == Stop) 
                break;

            result.Add(protein);
        }

        return result.ToArray();
    }
}

File: elixir/protein-translation/protein_translation.exs
defmodule ProteinTranslation do
  @doc """
  Given an RNA string, return a list of proteins specified by codons, in order.
  """
  @spec of_rna(String.t()) :: {atom, list(String.t())}
  def of_rna(rna) do
    { ok, peptides } = do_of_rna(rna, [])
    cond do
      ok == :error -> { :error, "invalid RNA" }
      true -> { :ok, peptides |> Enum.reverse }
    end
  end

  @proteins %{
    "UGU" => "Cysteine",
    "UGC" => "Cysteine",
    "UUA" => "Leucine",
    "UUG" => "Leucine",
    "AUG" => "Methionine",
    "UUU" => "Phenylalanine",
    "UUC" => "Phenylalanine",
    "UCU" => "Serine",
    "UCC" => "Serine",
    "UCA" => "Serine",
    "UCG" => "Serine",
    "UGG" => "Tryptophan",
    "UAU" => "Tyrosine",
    "UAC" => "Tyrosine",
    "UAA" => "STOP",
    "UAG" => "STOP",
    "UGA" => "STOP"
  }

  @codons @proteins |> Map.keys

  @doc """
  Given a codon, return the corresponding protein

  UGU -> Cysteine
  UGC -> Cysteine
  UUA -> Leucine
  UUG -> Leucine
  AUG -> Methionine
  UUU -> Phenylalanine
  UUC -> Phenylalanine
  UCU -> Serine
  UCC -> Serine
  UCA -> Serine
  UCG -> Serine
  UGG -> Tryptophan
  UAU -> Tyrosine
  UAC -> Tyrosine
  UAA -> STOP
  UAG -> STOP
  UGA -> STOP
  """
  @spec of_codon(String.t()) :: {atom, String.t()}
  def of_codon(codon) when codon in @codons, do: { :ok, @proteins[codon] }
  def of_codon(_), do: { :error, "invalid codon" }

  defp translate_rna("", acc), do: { :ok, acc }
  defp translate_rna(strand, acc) do
    { codon, rest } = strand |> String.split_at(3)
    { ok, peptide } = of_codon(codon)
    cond do
      ok == :error -> { :error, "invalid codon" }
      peptide == "STOP" -> { :ok, acc }
      true -> translate_rna(rest, [peptide|acc])
    end
  end
end
